doc(contracts/core): disclose compromised balancer contract vulnerability (#549)
cowprotocol/docs: docs/cow-protocol/reference/contracts/core/README.mdx

From the perspective of the API, two orders with the same parameters and the same owner but different app data are two different valid orders.
It can happen that the order that is part of the final settlement is the one controlled by the adversary, especially if other order parameters can be changed to create the appearence of an inflated surplus.
In this case, the order surplus decreases as partner fees are taken from the surplus.

### Loss of funds if a compromised Balancer contract is deployed

> A malicious solver can steal the entire amount of a user's trade if the Balancer contract is compromised and the solver processes a vulnerable user order.

Reviewer: If we want to add this, maybe is better to not make it the opening phrase and more a conclusion. Its both scary, and still unclear until we read a bit about this. 

The title is fine, here is another alternative in case you like it more: Solver extraction on networks with a compromised Balancer vault

Not sure if its better or worse, but I just want to reasure early in this writeup that mainnet for example is not affected.

Author: clarified the extreme circumstances required for there to be an issue.

> When a user is trading on certain affected networks and the user initiates a trade through a third party endpoint submitting a order type that has not been supported since 2024, it is possible that a malicious solver could steal the entire input amount of a user's trade.

Reviewer 2: Updating to include the most recent findings.

```suggestion
If the settlement contract deployment points to an absent or compromised Balancer vault, user orders may be seriously affected.
```

> When a user is trading on certain affected networks and the user initiates a trade through a third party endpoint submitting a order type that has not been supported since 2024, it is possible that a malicious solver could steal the entire input amount of a user's trade.

Reviewer 3: let's make this change

Due to a vulnerability in the [profanity vanity address generator](https://medium.com/@rebryk/how-to-hack-a-vanity-address-generated-with-profanity-ffad61ecacd2), the deployer wallet for the Balancer Vault was compromised. This allows arbitrary actors to deploy any bytecode to the address that is normally used to deploy [Balancer's Core Vault contract](https://docs.balancer.fi/concepts/vault/).

In order to preserve a consistent settlement address on all chains, CoW Protocol uses the `CREATE2` opcode to deploy the settlement contract. Since the balancer vault address is supplied as an immutable constructor property, the Settlement contract contains a dependency upon this compromised contract.

The Balancer vault is only used in `GPv2Transfer.sol` as an alternative way for a user to receive or supply funds from the settlement contract when the user sets `buyTokenBalance` or `sellTokenBalance` to `BALANCE_INTERNAL`.

There are two general effects that this vulnerability can have:

> * If the compromised contract at the `Balancer` address reverts upon a call to `manageUserBalance`, the attacker can effectively DoS any trade batches that contain a trade using `BALANCE_INTERNAL`.

Reviewer: ubernit: I personally would use "prevent users from trading" instead of using "DoS"

> There are two general effects that this vulnerability can have:
> * If the compromised contract at the `Balancer` address reverts upon a call to `manageUserBalance`, the attacker can effectively prevent users from trading in any trade batches that contain a trade using `BALANCE_INTERNAL`.
> * If the compromised contract does *not* revert, the settlement contract assumes that the user's balance was successfully transferred, and no further action is taken. This assumption allows for a solver to steal the full amount of a user's trade output, since the settlement contract does not verify that the funds were received by the user.

Reviewer 2: Updating to include the most recent findings. I wouldn't bother mentioning "user can't trade" because there's no Balancer vault there's no reason for the user to want to trade through it.

```suggestion
There are two levels for this vulnerability:
1. The target vault is specifically designed to exploit CoW user orders. 
2. The vault is empty or uses a contract that isn't designed to perform this specific exploit.

In case 1, the risks for the users are:
- Malicious solvers can steal `feeAmount` from *any existing user order*.
- Malicious solvers can make an order appear filled despite it not having been executed. This is particularly bad for specific classes of smart-contract orders relying on the filled amount of an order in their internal logic, as for example the ETH flow.  
- All risks for case 2.

In case 2, the risks for the users are:
- If the user signs an order using Balancer as the balance source/target for the order (i.e., using `internal` or `external` for `buyTokenBalance` or `sellTokenBalance`) then a solver may be able to steal the full amount of a user's trade output: the settlement contract assumes that the user's balance was successfully transferred from/to the vault, and no further action is taken.
```

In total, there are 3 conditions that are needed to make this vulnerability harmful:

> 1. The user must be trading on a network with a compromised Balancer contract

Reviewer: Maybe list networks where this is not the case

Author: there is a list slightly below. added additional comment to look there.

> 2. The trade must have been signed with `buyTokenBalance` or `sellTokenBalance` as `BALANCE_INTERNAL`

Reviewer: We should give some indication how unlikely this is from happening. Check this https://cowservices.slack.com/archives/C0361CCTFD5/p1761152767010129 and https://docs.google.com/spreadsheets/d/109_mybe7ZUEivJWLdd8jETQd3dixKXt-Z0LVstbhvyI/edit?gid=0#gid=0 

There's no orders with fee since March of last year. And if the API is blocking this orders, we should mention how this is not part of the solver competition anymore (adding the date, and ideally the  pointer to the code that rejects the order).

> 3. A malicious solver must obtain the signed order outside of the usual order flow (the CoW Orderbook API rejects these order types) and steal the user's funds. For the attack to be profitable, the stolen amount must exceed the current bond amount (as of writing, 1.5m USDC).

Reviewer: Ok, you mention here. I would write it a bit differently 

```suggestion
3. Users's don't place these orders since March 2024 (see Dune query)
4. These orders are not accepted anymore in the API since XXX YYYY (see code), therefore, they can't be part of the solver competition. So users would need to facilitate these orders using unofficial means to the solver
5. The stolen amount must exceed the current bond amount (as of writing, 1.5m USDC)
```

> In total, there are 3 conditions that are needed to make this vulnerability harmful:
> 1. The user must be trading on a network with a compromised Balancer contract (see affected networks below).
> 2. The trade must have been signed with `buyTokenBalance` or `sellTokenBalance` as `BALANCE_INTERNAL`. As of writing, this order type has not been processed by the protocol since March 2024.
> 3. The malicious solver must obtain the signed order outside of the usual order flow (the CoW Orderbook API rejects these order types) and steal the user's funds. For the attack to be profitable, the stolen amount must exceed the current bond amount (as of writing, 1.5m USDC).
> 

Reviewer 2: Unfortunately 2 and 3 don't hold anymore in the general case. 1 is already mentioned before so I'd drop the paragraph.

```suggestion
```

> As of writing, these are the CoW-supported networks affected by this vulnerability:
> * Plasma

Reviewer 2: Updating with latest findings plus fix chains. Plasma is good.

```suggestion
As of this writing, there are no official settlement contract deployments on chains where the Balancer vault can be used to compromise user orders that don't use the Balancer flow (case 1).

In any case, note that CoW orders are always  with `feeAmount = 0`, which would further limit the impact of this issue.

However, the following networks are currently at risk of loss of funds for user order intents that explicitly request trading through the Balancer vault (case 2):
* Linea
* Lens

Note that trading through the Balancer vault is currently unsupported at the protocol level and compromised orders cannot be posted to the orderbook.
```

:::warning

> In order to avoid being affected by this vulnerability, we do not recommend signing orders with `buyTokenBalance` or `sellTokenBalance` set to `BALANCE_INTERNAL`.

Reviewer 2: `BALANCE_INTERNAL` isn't what the user sees when signing, but rather `internal`. Also, `external` is also bad.

```suggestion
In order to avoid being affected by this vulnerability, we recommend only signing orders with `buyTokenBalance` or `sellTokenBalance` set to `erc20`.
```
